doc/synchronization.md

```diff
@@ -0,0 +1,356 @@
+> Quotations are used for notes.
+
+> This document is work-in-progress.
+> Intentions of this effort and document are: to sum up the behavior
+> of Ruby in concurrent and parallel environment, initiate discussion,
+> identify problems in the document, find flaws in the Ruby
+> implementations if any, suggest what has to be enhanced in Ruby itself
+> and cooperate towards the goal in all implementations (using
+> `concurrent-ruby` as compatibility layer).
+
+# Synchronization
+
+This layer provides tools to write concurrent abstractions independent on any
+particular Ruby implementation. It is build on top of Ruby memory model which is
+also described here. `concurrent-ruby` abstractions are build using this layer.
+
+**Why?** Ruby is great expressive language, but it lacks in support for
+concurrent and parallel computation. It's hoped that this document will provide
+ground steps for Ruby to become as good in this area as in others.
+
+Without memory model and this layer it's very hard to write concurrent
+abstractions for Ruby. To write a proper concurrent abstraction it often means
+to reimplement it more than once for different Ruby runtimes, which is very
+time-consuming and error-prone.
+
+# Ruby memory model
+
+Ruby memory model is a framework allowing to reason about programs in concurrent
+and parallel environment. It allows to identify what is and what is not a [race
+condition](https://en.wikipedia.org/wiki/Race_condition). Memory model is also a
+contract: if a program is written without the race conditions it behaves in
+[sequential consistent](https://en.wikipedia.org/wiki/Sequential_consistency)
+manner.
+
+Sources:
+
+-   [Java memory model](http://www.cs.umd.edu/~pugh/java/memoryModel/),
+    and its [FAQ](http://www.cs.umd.edu/~pugh/java/memoryModel/jsr-133-faq.html)
+-   [atomic&lt;&gt; Weapons 1](https://channel9.msdn.com/Shows/Going+Deep/Cpp-and-Beyond-2012-Herb-Sutter-atomic-Weapons-1-of-2)
+and
+[2](https://channel9.msdn.com/Shows/Going+Deep/Cpp-and-Beyond-2012-Herb-Sutter-atomic-Weapons-2-of-2)
+- [JRuby's wiki page about concurrency](https://github.com/jruby/jruby/wiki/Concurrency-in-jruby)
```

Comment:
I think we should think about how we talk about JRuby, Java and the JMM. A concern from casual readers could be that we are putting the JMM into Ruby or forcing the JRuby way of doing things onto MRI. How do we handle this question?